pkg/internal/controller/controller.go

```diff
@@ -173,10 +177,18 @@ func (c *Controller) Start(stop <-chan struct{}) error {
 func (c *Controller) processNextWorkItem() bool {
 	// This code copy-pasted from the sample-Controller.
 
+	// Create a context used for tagging OpenCensus measurements.
+	metricsCtx, _ := tag.New(context.Background(),
```

Comment:
If there are tags present in the background context, then those values will continue to be propagated to the [newly created context](https://sourcegraph.com/github.com/census-instrumentation/opencensus-go/-/blob/tag/map.go#L152-155). As such, if we use `tag.Insert` in subsequent calls for existing tags, the values existing tag keys [will not be updated](https://sourcegraph.com/github.com/census-instrumentation/opencensus-go/-/blob/tag/map.go#L95-98). 

Is that desired behavior?

Reply:
@logicalhan Are you saying the context returned by `context.Background()` may have tags present? I'm not aware of a way to add values to the background context, so I don't expect there will ever be tags present in it. As such I believe that in this code there is no effective difference between `tag.Insert` and `tag.Upsert`.

If I'm misunderstanding your question, please let me know. :)

Reply:
> To maintain contextual information of the specific request, tag propagation through your distributed system is necessary; the higher levels generate tags that are then passed down to the lower-level services. Data is collected with the tags and are sent to your observability systems.

>Above, a request comes in to the Web server. Web server tags all the outgoing requests with the following tags.
> originator=photo-app
> frontend=web
> These values are propagated all the way down to database and the CDN.

> With these tags, you can uniquely identify and break down which service called the downstream services, how much quota they’ve been used, what calls are failing and more.

In OpenCensus, tags can be propagated downstream and the values are stored in the context. Since this is the first time OpenCensus is being introduced, this is likely safe (since nothing else would be passing along OpenCensus tag information). However, that assumption may not hold in the future. 

I'm actually not saying that `tag.Insert` is the incorrect thing to do here. I'm just pointing out that there is a difference in behavior between `Insert` and `Update` and depending on what we actually want to do here, one would be more appropriate than the other. `Insert` can no-op __*if*__ we pass tag information to whatever triggers the controller. That may or may not be desirable.

Reply:
Actually, `Upsert` is probably safer, since it effectively forces the same behavior regardless of what happens upstream.

Reply:
Thanks, that clarifies things!

In this hypothetical future when the metrics context is received from an external source that may have added tags of its own, I believe `tag.Insert` is correct here. If the tag is already specified by the user of the library, the library should respect the user's intent and avoid overwriting their metrics tags with `tag.Upsert`.

Does that seem reasonable?

Reply:
I would imagine you would need to `Upsert` the result tag. If a request has gotten here, then it would presumably have been successful upstream.

Reply:
In the err'd cases.

Reply:
Actually all the result tags, you probably want to upsert.

Reply:
Since controllers can requeue things and pick them up again later, if something gets requeued, then wouldn't we expect the context have tags? It seems that we really do want to be judicious about using `Insert` or `Upsert`, since it will change existing outputted metric data. 

It may be worth adding a testcase around this by forcing a requeue and then forcing an error when it gets picked up again and comparing the measurements with what we'd expect.